edgelet/edgelet-http/src/certificate_manager.rs

```diff
@@ -0,0 +1,173 @@
+use std::str;
+use std::sync::{Arc, RwLock};
+
+use edgelet_core::crypto::{Certificate, CreateCertificate};
+use edgelet_core::{CertificateIssuer, CertificateProperties, CertificateType};
+use failure::ResultExt;
+
+pub use crate::error::{Error, ErrorKind};
+
+const IOTEDGED_VALIDITY: u64 = 7_776_000; // 90 days
+const IOTEDGED_TLS_COMMONNAME: &str = "iotedge tls";
+
+#[derive(Clone)]
+pub struct CertificateManager<C: CreateCertificate + Clone> {
+    certificate: Arc<RwLock<Option<String>>>,
+    crypto: C,
+}
+
+impl<C: CreateCertificate + Clone> CertificateManager<C> {
+    pub fn new(crypto_struct: C) -> Result<Self, Error> {
+        Ok(CertificateManager {
+            certificate: Arc::new(RwLock::new(None)),
+            crypto: crypto_struct,
+        })
+    }
+
+    pub fn get_certificate(&self) -> Result<String, Error> {
+        // First, try to directly read
+        {
+            let cert = self
+                .certificate
+                .read()
+                .expect("Locking the certificate for read failed.");
+
+            match Option::as_ref(&cert) {
+                Some(cert) => return Ok(cert.to_string()),
+                None => (),
+            };
+        }
+
+        // No valid cert so must create
+        let mut cert = self
+            .certificate
+            .write()
+            .expect("Locking the certificate for write failed.");
+
+        // Check that another thread hasn't already created one for us
+        match Option::as_ref(&cert) {
+            Some(cert) => Ok(cert.to_string()),
+            None => {
+                let new_cert = self
+                    .create_cert()
+                    .with_context(|_| ErrorKind::CertificateCreationError)?;
+                Ok(cert.get_or_insert(new_cert).to_string())
+            }
+        }
+    }
+
+    fn create_cert(&self) -> Result<String, Error> {
+        let edgelet_cert_props = CertificateProperties::new(
+            IOTEDGED_VALIDITY,
+            IOTEDGED_TLS_COMMONNAME.to_string(),
+            CertificateType::Server,
+            "iotedge-tls".to_string(),
```

Comment:
Should this be a literal string or an input? Not sure what the "alias" name in the cert is used for. Might be useful to check with Mahesh.